
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <title>What apps are there? &#8212; cogent3 2020.2.7a documentation</title>
    <link rel="stylesheet" href="../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/gallery.css" />
    <link rel="stylesheet" type="text/css" href="../_static/gallery-binder.css" />
    <link rel="stylesheet" type="text/css" href="../_static/gallery-dataframe.css" />
    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/language_data.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="../_static/require.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    <script src="https://unpkg.com/@jupyter-widgets/html-manager@^0.18.0/dist/embed-amd.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Using a nucleotide model" href="align-nucleotide.html" />
    <link rel="prev" title="Tracking records that could not be processed" href="not-completed.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  <div class="document">
    
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">cogent3</a></h1>








<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../warning.html">Warning</a></li>
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">The <code class="docutils literal notranslate"><span class="pre">apps</span></code></a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="index.html#overview">Overview</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#the-progressive-align-app">The <code class="docutils literal notranslate"><span class="pre">progressive_align</span></code> App</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#the-model-app">The <code class="docutils literal notranslate"><span class="pre">model</span></code> App</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#the-hypothesis-app">The <code class="docutils literal notranslate"><span class="pre">hypothesis</span></code> App</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#the-natsel-apps">The <code class="docutils literal notranslate"><span class="pre">natsel</span></code> Apps</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#the-user-function-apps">The <code class="docutils literal notranslate"><span class="pre">user_function</span></code> Apps</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../draw/index.html">Image Gallery</a></li>
<li class="toctree-l1"><a class="reference internal" href="../library.html">The Library</a></li>
<li class="toctree-l1"><a class="reference internal" href="../licenses.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="../general.html">Posting Bugs</a></li>
<li class="toctree-l1"><a class="reference internal" href="../general.html#citation">Citation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../general.html#support">Support</a></li>
<li class="toctree-l1"><a class="reference internal" href="../general.html#search">Search</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
  <li><a href="index.html">The <code class="docutils literal notranslate"><span class="pre">apps</span></code></a><ul>
      <li>Previous: <a href="not-completed.html" title="previous chapter">Tracking records that could not be processed</a></li>
      <li>Next: <a href="align-nucleotide.html" title="next chapter">Using a nucleotide model</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container,
div.nbinput.container div.prompt,
div.nbinput.container div.input_area,
div.nbinput.container div[class*=highlight],
div.nbinput.container div[class*=highlight] pre,
div.nboutput.container,
div.nboutput.container div.prompt,
div.nboutput.container div.output_area,
div.nboutput.container div[class*=highlight],
div.nboutput.container div[class*=highlight] pre {
    background: none;
    border: none;
    padding: 0 0;
    margin: 0;
    box-shadow: none;
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    min-width: 5.5ex;
    padding-top: 0.3rem;
    padding-right: 0.3rem;
    text-align: right;
    flex: 0;
}
@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    background: #f5f5f5;
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 0.3rem;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt a.copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}
</style>
<div class="section" id="What-apps-are-there?">
<h1>What apps are there?<a class="headerlink" href="#What-apps-are-there?" title="Permalink to this headline">¶</a></h1>
<p>Apps perform functions ranging from multiple sequence alignment (e.g. <code class="docutils literal notranslate"><span class="pre">progressive_align</span></code>), to excluding alignment columns containing non-nucleotide characters (e.g. <code class="docutils literal notranslate"><span class="pre">omit_degenerates</span></code>) to performing maximum-likelihood evolutionary analyses (e.g. <code class="docutils literal notranslate"><span class="pre">model</span></code>).</p>
<p>Apps that are identified as “composable” (with a value of <code class="docutils literal notranslate"><span class="pre">True</span></code> under that clumn) can be combined into a single function by addition.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">available_apps</span>

<span class="n">available_apps</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<table>
<style>
tr:last-child {border-bottom: 1px solid #000;} tr > th {text-align: center !important;} tr > td {text-align: left !important;}
</style>
<thead style="background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;">
<th>module</th>
<th>name</th>
<th>composable</th>
<th>doc</th>
<th>inputs</th>
<th>outputs</th>
<th>data type</th>
</thead>
<tbody>
<tr>
<td>cogent3.app.align</td>
<td>align_to_ref</td>
<td>True</td>
<td>Aligns to a reference seq, no gaps in the reference.
    Returns an Alignment object.</td>
<td>sequences</td>
<td>aligned, serialisable</td>
<td>SequenceCollection</td>
</tr>
<tr>
<td>cogent3.app.align</td>
<td>progressive_align</td>
<td>True</td>
<td>Progressive multiple sequence alignment via any cogent3 model.
     Returns an Alignment object.</td>
<td>sequences</td>
<td>aligned, serialisable</td>
<td>SequenceCollection</td>
</tr>
<tr>
<td>cogent3.app.dist</td>
<td>fast_slow_dist</td>
<td>True</td>
<td>Pairwise distance calculation. Uses fast (but less
    numerically robust) approach where possible, slow (robust)
    approach when not. Returns a DistanceMatrix.</td>
<td>aligned</td>
<td>pairwise_distances, serialisable</td>
<td>ArrayAlignment, Alignment</td>
</tr>
<tr>
<td>cogent3.app.evo</td>
<td>ancestral_states</td>
<td>True</td>
<td>Computes ancestral state probabilities from a model result. Returns a dict
    with a DictArray for each node.</td>
<td>model_result</td>
<td>result, tabular_result, serialisable</td>
<td>model_result</td>
</tr>
<tr>
<td>cogent3.app.evo</td>
<td>bootstrap</td>
<td>True</td>
<td>Parametric bootstrap for a provided hypothesis. Returns a bootstrap_result.</td>
<td>aligned</td>
<td>result, bootstrap_result, serialisable</td>
<td>ArrayAlignment, Alignment</td>
</tr>
<tr>
<td>cogent3.app.evo</td>
<td>hypothesis</td>
<td>True</td>
<td>Specify a hypothesis through defining two models. Returns a
    hypothesis_result.</td>
<td>aligned, serialisable</td>
<td>result, hypothesis_result, serialisable</td>
<td>ArrayAlignment, Alignment</td>
</tr>
<tr>
<td>cogent3.app.evo</td>
<td>model</td>
<td>True</td>
<td>Define a substitution model + tree for maximum likelihood evaluation.
    Returns model_result.</td>
<td>aligned, serialisable</td>
<td>result, model_result, serialisable</td>
<td>ArrayAlignment, Alignment</td>
</tr>
<tr>
<td>cogent3.app.evo</td>
<td>natsel_neutral</td>
<td>True</td>
<td>Test of selective neutrality by assessing whether omega equals 1.
    Under the alternate, there is one omega for all branches and all sites.</td>
<td>aligned, serialisable</td>
<td>result, hypothesis_result, serialisable</td>
<td>ArrayAlignment, Alignment</td>
</tr>
<tr>
<td>cogent3.app.evo</td>
<td>natsel_sitehet</td>
<td>True</td>
<td>Test for site-heterogeneity in omega. Under null, there are 2 site-classes,
    omega &lt; 1 and omega = 1. Under the alternate, an additional site-class of
    omega &gt; 1 is added.</td>
<td>aligned, serialisable</td>
<td>result, hypothesis_result, serialisable</td>
<td>ArrayAlignment, Alignment</td>
</tr>
<tr>
<td>cogent3.app.evo</td>
<td>natsel_timehet</td>
<td>True</td>
<td>The branch heterogeneity hypothesis test for natural selection.
    Tests for whether a single omega for all branches is sufficient against the
    alternate that a user specified subset of branches have a distinct value
    (or values) of omega.</td>
<td>aligned, serialisable</td>
<td>result, hypothesis_result, serialisable</td>
<td>ArrayAlignment, Alignment</td>
</tr>
<tr>
<td>cogent3.app.evo</td>
<td>natsel_zhang</td>
<td>True</td>
<td>The branch by site-class hypothesis test for natural selection of
    Zhang et al MBE 22: 2472-2479.

    Note: Our implementation is not as parametrically succinct as that of
    Zhang et al, we have 1 additional bin probability.</td>
<td>aligned, serialisable</td>
<td>result, hypothesis_result, serialisable</td>
<td>ArrayAlignment, Alignment</td>
</tr>
<tr>
<td>cogent3.app.evo</td>
<td>tabulate_stats</td>
<td>True</td>
<td>Extracts all model statistics from model_result as Table.</td>
<td>model_result, serialisable</td>
<td>result, tabular_result, serialisable</td>
<td>model_result</td>
</tr>
<tr>
<td>cogent3.app.io</td>
<td>load_aligned</td>
<td>True</td>
<td>Loads aligned sequences. Returns an Alignment object.</td>
<td></td>
<td>aligned, serialisable</td>
<td>DataStoreMember, str, Path</td>
</tr>
<tr>
<td>cogent3.app.io</td>
<td>load_db</td>
<td>True</td>
<td>Loads json serialised cogent3 objects from a TinyDB file.
    Returns whatever object type was stored.</td>
<td></td>
<td>serialisable</td>
<td></td>
</tr>
<tr>
<td>cogent3.app.io</td>
<td>load_json</td>
<td>True</td>
<td>Loads json serialised cogent3 objects from a json file.
    Returns whatever object type was stored.</td>
<td></td>
<td>serialisable</td>
<td></td>
</tr>
<tr>
<td>cogent3.app.io</td>
<td>load_tabular</td>
<td>True</td>
<td>Loads delimited data. Returns a Table.</td>
<td></td>
<td>tabular, serialisable</td>
<td>DataStoreMember, str, Path</td>
</tr>
<tr>
<td>cogent3.app.io</td>
<td>load_unaligned</td>
<td>True</td>
<td>Loads unaligned sequences. Returns a SequenceCollection.</td>
<td></td>
<td>sequences, serialisable</td>
<td>DataStoreMember, str, Path</td>
</tr>
<tr>
<td>cogent3.app.io</td>
<td>write_db</td>
<td>True</td>
<td>Writes json serialised objects to a TinyDB instance.</td>
<td>serialisable</td>
<td>identifier, serialisable</td>
<td></td>
</tr>
<tr>
<td>cogent3.app.io</td>
<td>write_json</td>
<td>True</td>
<td>Writes json serialised objects to individual json files.</td>
<td>serialisable</td>
<td>identifier, serialisable</td>
<td></td>
</tr>
<tr>
<td>cogent3.app.io</td>
<td>write_seqs</td>
<td>True</td>
<td>Writes sequences to text files in standard format.</td>
<td>sequences, aligned</td>
<td>sequences, aligned, identifier</td>
<td>ArrayAlignment, Alignment, SequenceCollection</td>
</tr>
<tr>
<td>cogent3.app.io</td>
<td>write_tabular</td>
<td>True</td>
<td>writes tabular data</td>
<td>tabular_result, tabular</td>
<td>identifier</td>
<td>Table, DictArray, DistanceMatrix</td>
</tr>
<tr>
<td>cogent3.app.sample</td>
<td>concat</td>
<td>False</td>
<td>Creates a concatenated alignment from a series. Returns an Alignment.</td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td>cogent3.app.sample</td>
<td>fixed_length</td>
<td>True</td>
<td>Sample an alignment to a fixed length. Returns an Alignment of the
    specified length, or NotCompleted if alignment too short.</td>
<td>aligned, serialisable</td>
<td>aligned, serialisable</td>
<td>ArrayAlignment, Alignment</td>
</tr>
<tr>
<td>cogent3.app.sample</td>
<td>min_length</td>
<td>True</td>
<td>Filters sequence collections / alignments by length. Returns the
    data if it satisfies the condition, NotCompleted otherwise.</td>
<td>sequences, aligned, serialisable</td>
<td>sequences, aligned, serialisable</td>
<td>ArrayAlignment, Alignment, SequenceCollection</td>
</tr>
<tr>
<td>cogent3.app.sample</td>
<td>omit_bad_seqs</td>
<td>True</td>
<td>Eliminates sequences from Alignment based on gap fraction, unique gaps.
    Returns modified alignment.</td>
<td>aligned, serialisable</td>
<td>aligned, serialisable</td>
<td>ArrayAlignment, Alignment</td>
</tr>
<tr>
<td>cogent3.app.sample</td>
<td>omit_degenerates</td>
<td>True</td>
<td>Excludes alignment columns with degenerate conditions. Can accomodate
    reading frame. Returns modified Alignment.</td>
<td>aligned, serialisable</td>
<td>aligned, serialisable</td>
<td>ArrayAlignment, Alignment</td>
</tr>
<tr>
<td>cogent3.app.sample</td>
<td>omit_duplicated</td>
<td>True</td>
<td>Removes redundant sequences, recording dropped sequences in
    seqs.info.dropped. Returns sequence collection with only unique sequences.</td>
<td>sequences, serialisable, aligned</td>
<td>sequences, serialisable, aligned</td>
<td>ArrayAlignment, Alignment, SequenceCollection</td>
</tr>
<tr>
<td>cogent3.app.sample</td>
<td>omit_gap_pos</td>
<td>True</td>
<td>Excludes gapped alignment columns meeting a threshold. Can accomodate
    reading frame. Returns modified Alignment.</td>
<td>aligned, serialisable</td>
<td>aligned, serialisable</td>
<td>ArrayAlignment, Alignment</td>
</tr>
<tr>
<td>cogent3.app.sample</td>
<td>take_codon_positions</td>
<td>True</td>
<td>Extracts the specified codon position(s) from an alignment.
    Returns an Alignment.</td>
<td>aligned, serialisable</td>
<td>aligned, serialisable</td>
<td>ArrayAlignment, Alignment</td>
</tr>
<tr>
<td>cogent3.app.sample</td>
<td>take_named_seqs</td>
<td>True</td>
<td>Extracts (or everything but) named sequences. Returns a filtered
    sequences, alignment that satisified the condition, NotCompleted otherwise.</td>
<td>sequences, aligned, serialisable</td>
<td>sequences, aligned, serialisable</td>
<td>ArrayAlignment, Alignment, SequenceCollection</td>
</tr>
<tr>
<td>cogent3.app.sample</td>
<td>trim_stop_codons</td>
<td>True</td>
<td>Removes terminal stop codons. Returns sequences / alignment.</td>
<td>sequences, aligned, serialisable</td>
<td>sequences, aligned, serialisable</td>
<td>ArrayAlignment, Alignment, SequenceCollection</td>
</tr>
<tr>
<td>cogent3.app.translate</td>
<td>select_translatable</td>
<td>True</td>
<td>Identifies most likely reading frame. Returns modified sequences / alignment,
    if it could be resolved, NotCompleted otherwise.</td>
<td>sequences, aligned</td>
<td>sequences</td>
<td>ArrayAlignment, Alignment, SequenceCollection</td>
</tr>
<tr>
<td>cogent3.app.translate</td>
<td>translate_seqs</td>
<td>True</td>
<td>Translates sequences, assumes in correct reading frame.</td>
<td>sequences, aligned</td>
<td>sequences, aligned</td>
<td>ArrayAlignment, Alignment, SequenceCollection</td>
</tr>
<tr>
<td>cogent3.app.tree</td>
<td>quick_tree</td>
<td>True</td>
<td>Neighbour Joining tree based on pairwise distances.</td>
<td>pairwise_distances</td>
<td>tree, serialisable</td>
<td>DistanceMatrix</td>
</tr>
<tr>
<td>cogent3.app.tree</td>
<td>scale_branches</td>
<td>True</td>
<td>Transforms tree branch lengths from nucleotide to codon, or the converse.
    Returns a Tree.</td>
<td>tree</td>
<td>tree, serialisable</td>
<td>PhyloNode, TreeNode</td>
</tr>
<tr>
<td>cogent3.app.tree</td>
<td>uniformize_tree</td>
<td>True</td>
<td>Standardises the orientation of unrooted trees. Returns a Tree.</td>
<td>tree</td>
<td>tree, serialisable</td>
<td>PhyloNode, TreeNode</td>
</tr>
<tr>
<td>cogent3.app.composable</td>
<td>user_function</td>
<td>True</td>
<td>wrapper class for user specified function</td>
<td></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<p>
37 rows x 7 columns</p></div>
</div>
<script type="application/vnd.jupyter.widget-state+json">
{"state": {}, "version_major": 2, "version_minor": 0}
</script></div>


          </div>
          
        </div>
      </div>
    <div class="clearer"></div>
  </div>
    <div class="footer">
      &copy;2020, cogent3.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 3.0.2</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="../_sources/app/available-apps.ipynb.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>