
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <title>Use an empirical protein substitution model &#8212; cogent3 2020.2.7a documentation</title>
    <link rel="stylesheet" href="../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/gallery.css" />
    <link rel="stylesheet" type="text/css" href="../_static/gallery-binder.css" />
    <link rel="stylesheet" type="text/css" href="../_static/gallery-dataframe.css" />
    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/language_data.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="../_static/require.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    <script src="https://unpkg.com/@jupyter-widgets/html-manager@^0.18.0/dist/embed-amd.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Likelihood analysis of multiple loci" href="testing_multi_loci.html" />
    <link rel="prev" title="Using codon models" href="codon_models.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  <div class="document">
    
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">cogent3</a></h1>








<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../warning.html">Warning</a></li>
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../app/index.html">The <code class="docutils literal notranslate"><span class="pre">apps</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../draw/index.html">Image Gallery</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../library.html">The Library</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../cookbook/index.html">Cookbook</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="index.html">Usage Examples</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/index.html">API</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_file_links.html">Data Files Used in the Documentation</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../licenses.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="../general.html">Posting Bugs</a></li>
<li class="toctree-l1"><a class="reference internal" href="../general.html#citation">Citation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../general.html#support">Support</a></li>
<li class="toctree-l1"><a class="reference internal" href="../general.html#search">Search</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
  <li><a href="../library.html">The Library</a><ul>
  <li><a href="index.html">Usage Examples</a><ul>
      <li>Previous: <a href="codon_models.html" title="previous chapter">Using codon models</a></li>
      <li>Next: <a href="testing_multi_loci.html" title="next chapter">Likelihood analysis of multiple loci</a></li>
  </ul></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="use-an-empirical-protein-substitution-model">
<h1>Use an empirical protein substitution model<a class="headerlink" href="#use-an-empirical-protein-substitution-model" title="Permalink to this headline">¶</a></h1>
<p><em>Section author: Gavin Huttley</em></p>
<p>This file contains an example of importing an empirically determined protein substitution matrix such as Dayhoff et al 1978 and using it to create a substitution model. The globin alignment is from the PAML distribution.</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_aligned_seqs</span><span class="p">,</span> <span class="n">make_tree</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent3.evolve.substitution_model</span> <span class="kn">import</span> <span class="n">EmpiricalProteinMatrix</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent3.parse.paml_matrix</span> <span class="kn">import</span> <span class="n">PamlMatrixParser</span>
</pre></div>
</div>
<p>Make a tree object.  In this case from a string.</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">treestring</span><span class="o">=</span><span class="s2">&quot;(((rabbit,rat),human),goat-cow,marsupial);&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span> <span class="o">=</span> <span class="n">make_tree</span><span class="p">(</span><span class="n">treestring</span><span class="p">)</span>
</pre></div>
</div>
<p>Import the alignment, explicitly setting the <code class="docutils literal notranslate"><span class="pre">moltype</span></code> to be protein</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">al</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span><span class="s1">&#39;data/abglobin_aa.phylip&#39;</span><span class="p">,</span>
<span class="gp">... </span>               <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;protein&quot;</span><span class="p">,</span>
<span class="gp">... </span>               <span class="p">)</span>
</pre></div>
</div>
<p>Open the file that contains the empirical matrix and parse the matrix and frequencies.</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">matrix_file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;data/dayhoff.dat&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">PamlMatrixParser</span></code> will import the matrix and frequency from files designed for Yang’s PAML package.  This format is the lower half of the matrix in three letter amino acid name order white space delineated followed by motif frequencies in the same order.</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">empirical_matrix</span><span class="p">,</span> <span class="n">empirical_frequencies</span> <span class="o">=</span> <span class="n">PamlMatrixParser</span><span class="p">(</span><span class="n">matrix_file</span><span class="p">)</span>
</pre></div>
</div>
<p>Create an Empirical Protein Matrix Substitution model object.  This will take the unscaled empirical matrix and use it and the motif frequencies to create a scaled Q matrix.</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sm</span> <span class="o">=</span> <span class="n">EmpiricalProteinMatrix</span><span class="p">(</span><span class="n">empirical_matrix</span><span class="p">,</span> <span class="n">empirical_frequencies</span><span class="p">)</span>
</pre></div>
</div>
<p>Make a parameter controller, likelihood function object and optimise.</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span> <span class="o">=</span> <span class="n">sm</span><span class="o">.</span><span class="n">make_likelihood_function</span><span class="p">(</span><span class="n">t</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span><span class="o">.</span><span class="n">set_alignment</span><span class="p">(</span><span class="n">al</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span><span class="o">.</span><span class="n">optimise</span><span class="p">(</span><span class="n">show_progress</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">lf</span><span class="o">.</span><span class="n">get_log_likelihood</span><span class="p">())</span>
<span class="go">-1706...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">lf</span><span class="p">)</span>
<span class="go">Likelihood function statistics</span>
<span class="go">log-likelihood = -1706.2489</span>
<span class="go">number of free parameters = 7</span>
<span class="go">=============================</span>
<span class="go">     edge    parent    length</span>
<span class="go">-----------------------------</span>
<span class="go">   rabbit    edge.0    0.0785</span>
<span class="go">      rat    edge.0    0.1750</span>
<span class="go">   edge.0    edge.1    0.0324</span>
<span class="go">    human    edge.1    0.0545</span>
<span class="go">   edge.1      root    0.0269</span>
<span class="go"> goat-cow      root    0.0972</span>
<span class="go">marsupial      root    0.2424</span>
<span class="go">-----------------------------</span>
<span class="go">============================================================================</span>
<span class="go">     A         C         D         E         F         G         H         I</span>
<span class="go">----------------------------------------------------------------------------</span>
<span class="go">0.0871    0.0335    0.0469    0.0495    0.0398    0.0886    0.0336    0.0369</span>
<span class="go">----------------------------------------------------------------------------...</span>
</pre></div>
</div>
</div>


          </div>
          
        </div>
      </div>
    <div class="clearer"></div>
  </div>
    <div class="footer">
      &copy;2020, cogent3.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 3.0.2</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="../_sources/examples/empirical_protein_models.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>