
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <title>Allowing substitution model parameters to differ between branches &#8212; cogent3 2020.2.7a documentation</title>
    <link rel="stylesheet" href="../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/gallery.css" />
    <link rel="stylesheet" type="text/css" href="../_static/gallery-binder.css" />
    <link rel="stylesheet" type="text/css" href="../_static/gallery-dataframe.css" />
    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/language_data.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="../_static/require.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    <script src="https://unpkg.com/@jupyter-widgets/html-manager@^0.18.0/dist/embed-amd.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Using codon models" href="codon_models.html" />
    <link rel="prev" title="A test of the neutral theory" href="neutral_test.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  <div class="document">
    
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">cogent3</a></h1>








<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../warning.html">Warning</a></li>
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../app/index.html">The <code class="docutils literal notranslate"><span class="pre">apps</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../draw/index.html">Image Gallery</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../library.html">The Library</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../cookbook/index.html">Cookbook</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="index.html">Usage Examples</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/index.html">API</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_file_links.html">Data Files Used in the Documentation</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../licenses.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="../general.html">Posting Bugs</a></li>
<li class="toctree-l1"><a class="reference internal" href="../general.html#citation">Citation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../general.html#support">Support</a></li>
<li class="toctree-l1"><a class="reference internal" href="../general.html#search">Search</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
  <li><a href="../library.html">The Library</a><ul>
  <li><a href="index.html">Usage Examples</a><ul>
      <li>Previous: <a href="neutral_test.html" title="previous chapter">A test of the neutral theory</a></li>
      <li>Next: <a href="codon_models.html" title="next chapter">Using codon models</a></li>
  </ul></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="allowing-substitution-model-parameters-to-differ-between-branches">
<span id="scope-params-on-trees"></span><h1>Allowing substitution model parameters to differ between branches<a class="headerlink" href="#allowing-substitution-model-parameters-to-differ-between-branches" title="Permalink to this headline">¶</a></h1>
<p><em>Section author: Gavin Huttley</em></p>
<p>A common task concerns assessing how substitution model exchangeability parameters differ between evolutionary lineages. This is most commonly of interest for the case of testing for natural selection. Here I’ll demonstrate the different ways of scoping parameters across trees for the codon model case and how these can be used for evolutionary modelling.</p>
<p>We start with the standard imports, plus using a canned codon substitution model and then load the sample data set.</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_aligned_seqs</span><span class="p">,</span> <span class="n">load_tree</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent3.evolve.models</span> <span class="kn">import</span> <span class="n">MG94HKY</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aln</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span><span class="s2">&quot;data/long_testseqs.fasta&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tree</span> <span class="o">=</span> <span class="n">load_tree</span><span class="p">(</span><span class="s2">&quot;data/test.tree&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>We construct the substitution model and likelihood function and set the alignment.</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sm</span> <span class="o">=</span> <span class="n">MG94HKY</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span> <span class="o">=</span> <span class="n">sm</span><span class="o">.</span><span class="n">make_likelihood_function</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span> <span class="n">digits</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">space</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span><span class="o">.</span><span class="n">set_alignment</span><span class="p">(</span><span class="n">aln</span><span class="p">)</span>
</pre></div>
</div>
<p>At this point we have a likelihood function with two exchangeability parameters from the substitution model (<code class="docutils literal notranslate"><span class="pre">kappa</span></code> the transition/transversion ratio; <code class="docutils literal notranslate"><span class="pre">omega</span></code> the nonsynonymous/synonymous ratio) plus branch lengths for all tree edges. To facilitate subsequent discussion I now display the tree</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">ascii_art</span><span class="p">())</span>
<span class="go">                              /-Human</span>
<span class="go">                    /edge.0--|</span>
<span class="go">          /edge.1--|          \-HowlerMon</span>
<span class="go">         |         |</span>
<span class="go">         |          \-Mouse</span>
<span class="go">-root----|</span>
<span class="go">         |--NineBande</span>
<span class="go">         |</span>
<span class="go">          \-DogFaced</span>
</pre></div>
</div>
<p>In order to scope a parameter on a tree (meaning specifying a subset of edges for which the parameter is to be treated differently to the remainder of the tree) requires uniquely identifying the edges. We do this using the following arguments to the likelihood function <code class="docutils literal notranslate"><span class="pre">set_param_rule</span></code> method:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">tip_names</span></code>: the name of two tips</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">outgroup_name</span></code>: the name of a tip that is not part of the clade of interest</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">clade</span></code>: if <code class="docutils literal notranslate"><span class="pre">True</span></code>, all lineages descended from the tree node identified by the <code class="docutils literal notranslate"><span class="pre">tip_names</span></code> and <code class="docutils literal notranslate"><span class="pre">outgroup_name</span></code> argument are affected by the other arguments. If <code class="docutils literal notranslate"><span class="pre">False</span></code>, then the <code class="docutils literal notranslate"><span class="pre">stem</span></code> argument must apply.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">stem</span></code>: Whether the edge leading to the node is included.</p></li>
</ul>
<p>The next concepts include exactly what can be scoped and how. In the case of testing for distinctive periods of natural selection it is common to specify distinct values for <code class="docutils literal notranslate"><span class="pre">omega</span></code> for an edge. I’ll first illustrate some possible uses for the arguments above by setting <code class="docutils literal notranslate"><span class="pre">omega</span></code> to be distinctive for specific edges. I will set a value for omega so that printing the likelihood function illustrates what edges have been effected, but I won’t actually do any model fitting.</p>
<div class="section" id="specifying-a-clade">
<h2>Specifying a clade<a class="headerlink" href="#specifying-a-clade" title="Permalink to this headline">¶</a></h2>
<p>I’m going to cause <code class="docutils literal notranslate"><span class="pre">omega</span></code> to attain a different value for all branches aside from the primate clade and stem (<code class="docutils literal notranslate"><span class="pre">HowlerMon</span></code>, <code class="docutils literal notranslate"><span class="pre">Human</span></code>, <code class="docutils literal notranslate"><span class="pre">edge.0</span></code>).</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span><span class="o">.</span><span class="n">set_param_rule</span><span class="p">(</span><span class="s1">&#39;omega&#39;</span><span class="p">,</span> <span class="n">tip_names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;DogFaced&#39;</span><span class="p">,</span> <span class="s1">&#39;Mouse&#39;</span><span class="p">],</span>
<span class="gp">... </span>             <span class="n">outgroup_name</span><span class="o">=</span><span class="s1">&#39;Human&#39;</span><span class="p">,</span> <span class="n">init</span><span class="o">=</span><span class="mf">2.0</span><span class="p">,</span> <span class="n">clade</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">lf</span><span class="p">)</span>
<span class="go">Likelihood function statistics</span>
<span class="go">log-likelihood = -9489.9506</span>
<span class="go">number of free parameters = 10</span>
<span class="go">=====</span>
<span class="go">kappa</span>
<span class="go">-----</span>
<span class="go"> 1.00</span>
<span class="go">-----</span>
<span class="go">===================================</span>
<span class="go">     edge   parent   length   omega</span>
<span class="go">-----------------------------------</span>
<span class="go">    Human   edge.0     0.03    1.00</span>
<span class="go">HowlerMon   edge.0     0.04    1.00</span>
<span class="go">   edge.0   edge.1     0.04    1.00</span>
<span class="go">    Mouse   edge.1     0.28    2.00</span>
<span class="go">   edge.1     root     0.02    2.00</span>
<span class="go">NineBande     root     0.09    2.00</span>
<span class="go"> DogFaced     root     0.11    2.00</span>
<span class="go">-----------------------------------</span>
<span class="go">=========================</span>
<span class="go">   A      C      G      T</span>
<span class="go">-------------------------</span>
<span class="go">0.37   0.19   0.21   0.23</span>
<span class="go">-------------------------</span>
</pre></div>
</div>
<p>As you can see <code class="docutils literal notranslate"><span class="pre">omega</span></code> for the primate edges I listed above have the default parameter value (1.0), while the others have what I’ve assigned. In fact, you could omit the <code class="docutils literal notranslate"><span class="pre">clade</span></code> argument as this is the default, but I think for readability of scripts it’s best to be explicit.</p>
</div>
<div class="section" id="specifying-a-stem">
<h2>Specifying a stem<a class="headerlink" href="#specifying-a-stem" title="Permalink to this headline">¶</a></h2>
<p>This time I’ll specify the stem leading to the primates as the edge of interest.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>I need to reset the <code class="docutils literal notranslate"><span class="pre">lf</span></code> so all edges have the default value again. I’ll show this only for this example, but rest assured I’m doing it for all others too.</p>
</div>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span><span class="o">.</span><span class="n">set_param_rule</span><span class="p">(</span><span class="s1">&#39;omega&#39;</span><span class="p">,</span> <span class="n">init</span><span class="o">=</span><span class="mf">1.0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span><span class="o">.</span><span class="n">set_param_rule</span><span class="p">(</span><span class="s1">&#39;omega&#39;</span><span class="p">,</span> <span class="n">tip_names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Human&#39;</span><span class="p">,</span> <span class="s1">&#39;HowlerMon&#39;</span><span class="p">],</span>
<span class="gp">... </span>     <span class="n">outgroup_name</span><span class="o">=</span><span class="s1">&#39;Mouse&#39;</span><span class="p">,</span> <span class="n">init</span><span class="o">=</span><span class="mf">2.0</span><span class="p">,</span> <span class="n">stem</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">clade</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">lf</span><span class="p">)</span>
<span class="go">Likelihood function statistics</span>
<span class="go">log-likelihood = -9424.8896</span>
<span class="go">number of free parameters = 10</span>
<span class="go">=====</span>
<span class="go">kappa</span>
<span class="go">-----</span>
<span class="go"> 1.00</span>
<span class="go">-----</span>
<span class="go">===================================</span>
<span class="go">     edge   parent   length   omega</span>
<span class="go">-----------------------------------</span>
<span class="go">    Human   edge.0     0.03    1.00</span>
<span class="go">HowlerMon   edge.0     0.04    1.00</span>
<span class="go">   edge.0   edge.1     0.04    2.00</span>
<span class="go">    Mouse   edge.1     0.28    1.00</span>
<span class="go">   edge.1     root     0.02    1.00</span>
<span class="go">NineBande     root     0.09    1.00</span>
<span class="go"> DogFaced     root     0.11    1.00</span>
<span class="go">-----------------------------------...</span>
</pre></div>
</div>
</div>
<div class="section" id="specifying-clade-and-stem">
<h2>Specifying clade and stem<a class="headerlink" href="#specifying-clade-and-stem" title="Permalink to this headline">¶</a></h2>
<p>I’ll specify that both the primates and their stem are to be considered.</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span><span class="o">.</span><span class="n">set_param_rule</span><span class="p">(</span><span class="s1">&#39;omega&#39;</span><span class="p">,</span> <span class="n">tip_names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Human&#39;</span><span class="p">,</span> <span class="s1">&#39;HowlerMon&#39;</span><span class="p">],</span>
<span class="gp">... </span>     <span class="n">outgroup_name</span><span class="o">=</span><span class="s1">&#39;Mouse&#39;</span><span class="p">,</span> <span class="n">init</span><span class="o">=</span><span class="mf">2.0</span><span class="p">,</span> <span class="n">stem</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">clade</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">lf</span><span class="p">)</span>
<span class="go">Likelihood function statistics</span>
<span class="go">log-likelihood = -9442.4271</span>
<span class="go">number of free parameters = 10</span>
<span class="go">=====</span>
<span class="go">kappa</span>
<span class="go">-----</span>
<span class="go"> 1.00</span>
<span class="go">-----</span>
<span class="go">===================================</span>
<span class="go">     edge   parent   length   omega</span>
<span class="go">-----------------------------------</span>
<span class="go">    Human   edge.0     0.03    2.00</span>
<span class="go">HowlerMon   edge.0     0.04    2.00</span>
<span class="go">   edge.0   edge.1     0.04    2.00</span>
<span class="go">    Mouse   edge.1     0.28    1.00</span>
<span class="go">   edge.1     root     0.02    1.00</span>
<span class="go">NineBande     root     0.09    1.00</span>
<span class="go"> DogFaced     root     0.11    1.00</span>
<span class="go">-----------------------------------...</span>
</pre></div>
</div>
</div>
<div class="section" id="alternate-arguments-for-specifying-edges">
<h2>Alternate arguments for specifying edges<a class="headerlink" href="#alternate-arguments-for-specifying-edges" title="Permalink to this headline">¶</a></h2>
<p>The likelihood function <code class="docutils literal notranslate"><span class="pre">set_param_rule</span></code> method also has the arguments of <code class="docutils literal notranslate"><span class="pre">edge</span></code> and <code class="docutils literal notranslate"><span class="pre">edges</span></code>. These allow specific naming of the tree edge(s) to be affected by a rule. In general, however, the <code class="docutils literal notranslate"><span class="pre">tip_names</span></code> + <code class="docutils literal notranslate"><span class="pre">outgroup_name</span></code> combo is more robust.</p>
</div>
<div class="section" id="applications-of-scoped-parameters">
<h2>Applications of scoped parameters<a class="headerlink" href="#applications-of-scoped-parameters" title="Permalink to this headline">¶</a></h2>
<p>The general use-cases for which a tree scope can be applied are:</p>
<ol class="arabic">
<li><p>constraining all edges identified by a rule to have a specific value which is constant and not modifiable</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span><span class="o">.</span><span class="n">set_param_rule</span><span class="p">(</span><span class="s1">&#39;omega&#39;</span><span class="p">,</span> <span class="n">tip_names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Human&#39;</span><span class="p">,</span> <span class="s1">&#39;HowlerMon&#39;</span><span class="p">],</span>
<span class="gp">... </span>     <span class="n">outgroup_name</span><span class="o">=</span><span class="s1">&#39;Mouse&#39;</span><span class="p">,</span> <span class="n">clade</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">is_constant</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</li>
<li><p>all edges identified by a rule have the same but different value to the rest of the tree</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span><span class="o">.</span><span class="n">set_param_rule</span><span class="p">(</span><span class="s1">&#39;omega&#39;</span><span class="p">,</span> <span class="n">tip_names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Human&#39;</span><span class="p">,</span> <span class="s1">&#39;HowlerMon&#39;</span><span class="p">],</span>
<span class="gp">... </span>     <span class="n">outgroup_name</span><span class="o">=</span><span class="s1">&#39;Mouse&#39;</span><span class="p">,</span> <span class="n">clade</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</li>
<li><p>allowing all edges identified by a rule to have different values of the parameter with the remaining tree edges having the same value</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span><span class="o">.</span><span class="n">set_param_rule</span><span class="p">(</span><span class="s1">&#39;omega&#39;</span><span class="p">,</span> <span class="n">tip_names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Human&#39;</span><span class="p">,</span> <span class="s1">&#39;HowlerMon&#39;</span><span class="p">],</span>
<span class="gp">... </span>     <span class="n">outgroup_name</span><span class="o">=</span><span class="s1">&#39;Mouse&#39;</span><span class="p">,</span> <span class="n">clade</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">is_independent</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</li>
<li><p>allowing all edges to have a different value</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span><span class="o">.</span><span class="n">set_param_rule</span><span class="p">(</span><span class="s1">&#39;omega&#39;</span><span class="p">,</span> <span class="n">is_independent</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</li>
</ol>
<p>I’ll demonstrate these cases sequentially as they involve gradually increasing the degrees of freedom in the model. First we’ll constrain <code class="docutils literal notranslate"><span class="pre">omega</span></code> to equal 1 on the primate edges. I’ll then optimise the model.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>here I’m specifying a constant value for the parameter and so I <strong>must</strong> use the argument <code class="docutils literal notranslate"><span class="pre">value</span></code> to set it. This not to be confused with the argument <code class="docutils literal notranslate"><span class="pre">init</span></code> that is used for providing initial (starting) values for fitting.</p>
</div>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span><span class="o">.</span><span class="n">set_param_rule</span><span class="p">(</span><span class="s1">&#39;omega&#39;</span><span class="p">,</span> <span class="n">tip_names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Human&#39;</span><span class="p">,</span> <span class="s1">&#39;HowlerMon&#39;</span><span class="p">],</span>
<span class="gp">... </span>     <span class="n">outgroup_name</span><span class="o">=</span><span class="s1">&#39;Mouse&#39;</span><span class="p">,</span> <span class="n">clade</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">is_constant</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span><span class="o">.</span><span class="n">optimise</span><span class="p">(</span><span class="n">local</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">show_progress</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">lf</span><span class="p">)</span>
<span class="go">Likelihood function statistics</span>
<span class="go">log-likelihood = -8640.9290</span>
<span class="go">number of free parameters = 9</span>
<span class="go">=====</span>
<span class="go">kappa</span>
<span class="go">-----</span>
<span class="go"> 3.87</span>
<span class="go">-----</span>
<span class="go">===================================</span>
<span class="go">     edge   parent   length   omega</span>
<span class="go">-----------------------------------</span>
<span class="go">    Human   edge.0     0.09    1.00</span>
<span class="go">HowlerMon   edge.0     0.12    1.00</span>
<span class="go">   edge.0   edge.1     0.12    0.92</span>
<span class="go">    Mouse   edge.1     0.84    0.92</span>
<span class="go">   edge.1     root     0.06    0.92</span>
<span class="go">NineBande     root     0.28    0.92</span>
<span class="go"> DogFaced     root     0.34    0.92</span>
<span class="go">-----------------------------------</span>
<span class="go">=========================</span>
<span class="go">   A      C      G      T</span>
<span class="go">-------------------------</span>
<span class="go">0.37   0.19   0.21   0.23</span>
<span class="go">-------------------------</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">lf</span><span class="o">.</span><span class="n">lnL</span><span class="p">)</span>
<span class="go">-8640.9...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">lf</span><span class="o">.</span><span class="n">nfp</span><span class="p">)</span>
<span class="go">9</span>
</pre></div>
</div>
<p>I’ll now free up <code class="docutils literal notranslate"><span class="pre">omega</span></code> on the primate clade, but making it a single value shared by all primate lineages.</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span><span class="o">.</span><span class="n">set_param_rule</span><span class="p">(</span><span class="s1">&#39;omega&#39;</span><span class="p">,</span> <span class="n">tip_names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Human&#39;</span><span class="p">,</span> <span class="s1">&#39;HowlerMon&#39;</span><span class="p">],</span>
<span class="gp">... </span>     <span class="n">outgroup_name</span><span class="o">=</span><span class="s1">&#39;Mouse&#39;</span><span class="p">,</span> <span class="n">clade</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">is_constant</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span><span class="o">.</span><span class="n">optimise</span><span class="p">(</span><span class="n">local</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">show_progress</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">lf</span><span class="p">)</span>
<span class="go">Likelihood function statistics</span>
<span class="go">log-likelihood = -8639.7171</span>
<span class="go">number of free parameters = 10</span>
<span class="go">=====</span>
<span class="go">kappa</span>
<span class="go">-----</span>
<span class="go"> 3.85</span>
<span class="go">-----</span>
<span class="go">===================================</span>
<span class="go">     edge   parent   length   omega</span>
<span class="go">-----------------------------------</span>
<span class="go">    Human   edge.0     0.09    0.77</span>
<span class="go">HowlerMon   edge.0     0.12    0.77</span>
<span class="go">   edge.0   edge.1     0.12    0.92</span>
<span class="go">    Mouse   edge.1     0.84    0.92</span>
<span class="go">   edge.1     root     0.06    0.92</span>
<span class="go">NineBande     root     0.28    0.92</span>
<span class="go"> DogFaced     root     0.34    0.92</span>
<span class="go">-----------------------------------</span>
<span class="go">=========================</span>
<span class="go">   A      C      G      T</span>
<span class="go">-------------------------</span>
<span class="go">0.37   0.19   0.21   0.23</span>
<span class="go">-------------------------</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">lf</span><span class="o">.</span><span class="n">lnL</span><span class="p">)</span>
<span class="go">-8639.7...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">lf</span><span class="o">.</span><span class="n">nfp</span><span class="p">)</span>
<span class="go">10</span>
</pre></div>
</div>
<p>Finally I’ll allow all primate edges to have different values of <code class="docutils literal notranslate"><span class="pre">omega</span></code>.</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span><span class="o">.</span><span class="n">set_param_rule</span><span class="p">(</span><span class="s1">&#39;omega&#39;</span><span class="p">,</span> <span class="n">tip_names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Human&#39;</span><span class="p">,</span> <span class="s1">&#39;HowlerMon&#39;</span><span class="p">],</span>
<span class="gp">... </span>     <span class="n">outgroup_name</span><span class="o">=</span><span class="s1">&#39;Mouse&#39;</span><span class="p">,</span> <span class="n">clade</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">is_independent</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span><span class="o">.</span><span class="n">optimise</span><span class="p">(</span><span class="n">local</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">show_progress</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">lf</span><span class="p">)</span>
<span class="go">Likelihood function statistics</span>
<span class="go">log-likelihood = -8638.9572</span>
<span class="go">number of free parameters = 11</span>
<span class="go">=====</span>
<span class="go">kappa</span>
<span class="go">-----</span>
<span class="go"> 3.85</span>
<span class="go">-----</span>
<span class="go">===================================</span>
<span class="go">     edge   parent   length   omega</span>
<span class="go">-----------------------------------</span>
<span class="go">    Human   edge.0     0.09    0.59</span>
<span class="go">HowlerMon   edge.0     0.12    0.95</span>
<span class="go">   edge.0   edge.1     0.12    0.92</span>
<span class="go">    Mouse   edge.1     0.84    0.92</span>
<span class="go">   edge.1     root     0.06    0.92</span>
<span class="go">NineBande     root     0.28    0.92</span>
<span class="go"> DogFaced     root     0.34    0.92</span>
<span class="go">-----------------------------------</span>
<span class="go">=========================</span>
<span class="go">   A      C      G      T</span>
<span class="go">-------------------------</span>
<span class="go">0.37   0.19   0.21   0.23</span>
<span class="go">-------------------------</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">lf</span><span class="o">.</span><span class="n">lnL</span><span class="p">)</span>
<span class="go">-8638.9...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">lf</span><span class="o">.</span><span class="n">nfp</span><span class="p">)</span>
<span class="go">11</span>
</pre></div>
</div>
<p>We now allow <code class="docutils literal notranslate"><span class="pre">omega</span></code> to be different on all edges.</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span><span class="o">.</span><span class="n">set_param_rule</span><span class="p">(</span><span class="s1">&#39;omega&#39;</span><span class="p">,</span> <span class="n">is_independent</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lf</span><span class="o">.</span><span class="n">optimise</span><span class="p">(</span><span class="n">local</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">show_progress</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">lf</span><span class="p">)</span>
<span class="go">Likelihood function statistics</span>
<span class="go">log-likelihood = -8636.1383</span>
<span class="go">number of free parameters = 15</span>
<span class="go">=====</span>
<span class="go">kappa</span>
<span class="go">-----</span>
<span class="go"> 3.85</span>
<span class="go">-----</span>
<span class="go">===================================</span>
<span class="go">     edge   parent   length   omega</span>
<span class="go">-----------------------------------</span>
<span class="go">    Human   edge.0     0.09    0.59</span>
<span class="go">HowlerMon   edge.0     0.12    0.95</span>
<span class="go">   edge.0   edge.1     0.12    1.13</span>
<span class="go">    Mouse   edge.1     0.84    0.92</span>
<span class="go">   edge.1     root     0.06    0.38</span>
<span class="go">NineBande     root     0.28    1.27</span>
<span class="go"> DogFaced     root     0.34    0.84</span>
<span class="go">-----------------------------------</span>
<span class="go">=========================</span>
<span class="go">   A      C      G      T</span>
<span class="go">-------------------------</span>
<span class="go">0.37   0.19   0.21   0.23</span>
<span class="go">-------------------------</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">lf</span><span class="o">.</span><span class="n">lnL</span><span class="p">)</span>
<span class="go">-8636.1...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">lf</span><span class="o">.</span><span class="n">nfp</span><span class="p">)</span>
<span class="go">15</span>
</pre></div>
</div>
</div>
</div>


          </div>
          
        </div>
      </div>
    <div class="clearer"></div>
  </div>
    <div class="footer">
      &copy;2020, cogent3.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 3.0.2</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="../_sources/examples/scope_model_params_on_trees.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>