
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <title>Advanced sequence handling &#8212; cogent3 2020.2.7a documentation</title>
    <link rel="stylesheet" href="../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/gallery.css" />
    <link rel="stylesheet" type="text/css" href="../_static/gallery-binder.css" />
    <link rel="stylesheet" type="text/css" href="../_static/gallery-dataframe.css" />
    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/language_data.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="../_static/require.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    <script src="https://unpkg.com/@jupyter-widgets/html-manager@^0.18.0/dist/embed-amd.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Complete version of manipulating sequence annotations" href="complete_seq_features.html" />
    <link rel="prev" title="Translating DNA into protein" href="translate_dna.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  <div class="document">
    
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">cogent3</a></h1>








<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../warning.html">Warning</a></li>
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../app/index.html">The <code class="docutils literal notranslate"><span class="pre">apps</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../draw/index.html">Image Gallery</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../library.html">The Library</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../cookbook/index.html">Cookbook</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="index.html">Usage Examples</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/index.html">API</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_file_links.html">Data Files Used in the Documentation</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../licenses.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="../general.html">Posting Bugs</a></li>
<li class="toctree-l1"><a class="reference internal" href="../general.html#citation">Citation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../general.html#support">Support</a></li>
<li class="toctree-l1"><a class="reference internal" href="../general.html#search">Search</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
  <li><a href="../library.html">The Library</a><ul>
  <li><a href="index.html">Usage Examples</a><ul>
      <li>Previous: <a href="translate_dna.html" title="previous chapter">Translating DNA into protein</a></li>
      <li>Next: <a href="complete_seq_features.html" title="next chapter">Complete version of manipulating sequence annotations</a></li>
  </ul></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="advanced-sequence-handling">
<span id="seq-annotations"></span><h1>Advanced sequence handling<a class="headerlink" href="#advanced-sequence-handling" title="Permalink to this headline">¶</a></h1>
<p><em>Section author: Gavin Huttley</em></p>
<p>Individual sequences and alignments can be manipulated by annotations. Most value in the genome sequences arises from sequence annotations regarding specific sequence feature types, e.g. genes with introns / exons, repeat sequences. These can be applied to an alignment either using data formats available from genome portals (e.g. GFF, or GenBank annotation formats) or by custom assignments.</p>
<p>Annotations can be added in two ways: using either the <code class="docutils literal notranslate"><span class="pre">add_annotation</span></code> or the <code class="docutils literal notranslate"><span class="pre">add_feature</span></code> method. The distinction between these two is that <code class="docutils literal notranslate"><span class="pre">add_features</span></code> is more specialised. Features can be thought of as a type of annotation representing standard sequence properties eg introns/exons. Annotations are the more general case, such as a computed property which has, say a numerical value and a span.</p>
<p>For illustrative purposes we define a sequence with 2 exons and grab the 1<sup>st</sup> exon:</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">DNA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="n">DNA</span><span class="o">.</span><span class="n">make_seq</span><span class="p">(</span><span class="s2">&quot;aagaagaagacccccaaaaaaaaaattttttttttaaaaaaaaaaaaa&quot;</span><span class="p">,</span>
<span class="gp">... </span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;Orig&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">exon1</span> <span class="o">=</span> <span class="n">s</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s1">&#39;exon&#39;</span><span class="p">,</span> <span class="s1">&#39;exon1&#39;</span><span class="p">,</span> <span class="p">[(</span><span class="mi">10</span><span class="p">,</span><span class="mi">15</span><span class="p">)])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">exon2</span> <span class="o">=</span> <span class="n">s</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s1">&#39;exon&#39;</span><span class="p">,</span> <span class="s1">&#39;exon2&#39;</span><span class="p">,</span> <span class="p">[(</span><span class="mi">30</span><span class="p">,</span><span class="mi">40</span><span class="p">)])</span>
</pre></div>
</div>
<p>Here, ‘<code class="docutils literal notranslate"><span class="pre">exon</span></code>’ is the feature type, and ‘<code class="docutils literal notranslate"><span class="pre">exon#</span></code>’ the feature name. The feature type is used for the display formatting, which won’t be illustrated here, and also for selecting all features of the same type, shown below.</p>
<p>We could also have created an annotation using the <code class="docutils literal notranslate"><span class="pre">add_annotation</span></code> method:</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent3.core.annotation</span> <span class="kn">import</span> <span class="n">Feature</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s2</span><span class="o">=</span><span class="n">DNA</span><span class="o">.</span><span class="n">make_seq</span><span class="p">(</span><span class="s2">&quot;aagaagaagacccccaaaaaaaaaattttttttttaaaaaaaaaaaaa&quot;</span><span class="p">,</span>
<span class="gp">... </span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;Orig2&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">exon3</span> <span class="o">=</span> <span class="n">s2</span><span class="o">.</span><span class="n">add_annotation</span><span class="p">(</span><span class="n">Feature</span><span class="p">,</span> <span class="s1">&#39;exon&#39;</span><span class="p">,</span> <span class="s1">&#39;exon1&#39;</span><span class="p">,</span> <span class="p">[(</span><span class="mi">35</span><span class="p">,</span><span class="mi">40</span><span class="p">)])</span>
</pre></div>
</div>
<p>We can use the features (eg <code class="docutils literal notranslate"><span class="pre">exon1</span></code>) to get the corresponding sequence region.</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="p">[</span><span class="n">exon1</span><span class="p">]</span>
<span class="go">DnaSequence(CCCCC)</span>
</pre></div>
</div>
<p>You can query annotations by type and optionally by label, receiving a list of features:</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">exons</span> <span class="o">=</span> <span class="n">s</span><span class="o">.</span><span class="n">get_annotations_matching</span><span class="p">(</span><span class="s1">&#39;exon&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">exons</span><span class="p">)</span>
<span class="go">[exon &quot;exon1&quot; at [10:15]/48, exon &quot;exon2&quot; at [30:40]/48]</span>
</pre></div>
</div>
<p>We can use this list to construct a pseudo-feature covering (or excluding) multiple features using <code class="docutils literal notranslate"><span class="pre">get_region_covering_all</span></code>. For instance, getting all exons,</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">s</span><span class="o">.</span><span class="n">get_region_covering_all</span><span class="p">(</span><span class="n">exons</span><span class="p">))</span>
<span class="go">region &quot;exon&quot; at [10:15, 30:40]/48</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">.</span><span class="n">get_region_covering_all</span><span class="p">(</span><span class="n">exons</span><span class="p">)</span><span class="o">.</span><span class="n">get_slice</span><span class="p">()</span>
<span class="go">DnaSequence(CCCCCTT... 15)</span>
</pre></div>
</div>
<p>or not exons (the exon <em>shadow</em>):</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">s</span><span class="o">.</span><span class="n">get_region_covering_all</span><span class="p">(</span><span class="n">exons</span><span class="p">)</span><span class="o">.</span><span class="n">get_shadow</span><span class="p">()</span><span class="o">.</span><span class="n">get_slice</span><span class="p">())</span>
<span class="go">AAGAAGAAGAAAAAAAAAAATTTTTAAAAAAAA</span>
</pre></div>
</div>
<p>The first of these essentially returns the CDS of the gene.</p>
<p>Features are themselves sliceable:</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">exon1</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">get_slice</span><span class="p">()</span>
<span class="go">DnaSequence(CCC)</span>
</pre></div>
</div>
<p>This approach to sequence / alignment handling allows the user to manipulate them according to things they know about such as genes or repeat elements. Most of this annotation data can be obtained from genome portals.</p>
<p>The toolkit can perform standard sequence / alignment manipulations such as getting a subset of sequences or aligned columns, translating sequences, reading and writing standard formats.</p>
</div>


          </div>
          
        </div>
      </div>
    <div class="clearer"></div>
  </div>
    <div class="footer">
      &copy;2020, cogent3.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 3.0.2</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="../_sources/examples/seq_features.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>