
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <title>Building phylogenies &#8212; cogent3 2020.2.7a documentation</title>
    <link rel="stylesheet" href="../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/gallery.css" />
    <link rel="stylesheet" type="text/css" href="../_static/gallery-binder.css" />
    <link rel="stylesheet" type="text/css" href="../_static/gallery-dataframe.css" />
    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/language_data.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="../_static/require.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    <script src="https://unpkg.com/@jupyter-widgets/html-manager@^0.18.0/dist/embed-amd.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Evolutionary Analysis Using Likelihood" href="evo_modelling.html" />
    <link rel="prev" title="Genetic distance calculation" href="calc_genetic_distance.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  <div class="document">
    
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">cogent3</a></h1>








<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../warning.html">Warning</a></li>
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../app/index.html">The <code class="docutils literal notranslate"><span class="pre">apps</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../draw/index.html">Image Gallery</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../library.html">The Library</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="index.html">Cookbook</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/index.html">Usage Examples</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/index.html">API</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_file_links.html">Data Files Used in the Documentation</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../licenses.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="../general.html">Posting Bugs</a></li>
<li class="toctree-l1"><a class="reference internal" href="../general.html#citation">Citation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../general.html#support">Support</a></li>
<li class="toctree-l1"><a class="reference internal" href="../general.html#search">Search</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
  <li><a href="../library.html">The Library</a><ul>
  <li><a href="index.html">Cookbook</a><ul>
      <li>Previous: <a href="calc_genetic_distance.html" title="previous chapter">Genetic distance calculation</a></li>
      <li>Next: <a href="evo_modelling.html" title="next chapter">Evolutionary Analysis Using Likelihood</a></li>
  </ul></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="building-phylogenies">
<h1>Building phylogenies<a class="headerlink" href="#building-phylogenies" title="Permalink to this headline">¶</a></h1>
<div class="section" id="building-a-phylogenetic-tree-from-pairwise-distances">
<h2>Building A Phylogenetic Tree From Pairwise Distances<a class="headerlink" href="#building-a-phylogenetic-tree-from-pairwise-distances" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="directly-via-alignment-quick-tree">
<h2>Directly via <code class="docutils literal notranslate"><span class="pre">alignment.quick_tree()</span></code><a class="headerlink" href="#directly-via-alignment-quick-tree" title="Permalink to this headline">¶</a></h2>
<p>Both the <code class="docutils literal notranslate"><span class="pre">ArrayAlignment</span></code> and <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> classes support this.</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_aligned_seqs</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aln</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span><span class="s1">&#39;data/primate_brca1.fasta&#39;</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tree</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">quick_tree</span><span class="p">(</span><span class="n">calc</span><span class="o">=</span><span class="s2">&quot;TN93&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tree</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">balanced</span><span class="p">()</span>  <span class="c1"># purely for display</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">ascii_art</span><span class="p">())</span>  
<span class="go">                    /-Rhesus</span>
<span class="go">          /edge.1--|</span>
<span class="go">         |         |          /-HowlerMon</span>
<span class="go">         |          \edge.0--|</span>
<span class="go">         |                    \-Galago</span>
<span class="go">-root----|</span>
<span class="go">         |--Orangutan</span>
<span class="go">         |</span>
<span class="go">         |          /-Chimpanzee</span>
<span class="go">          \edge.2--|</span>
<span class="go">                   |          /-Human</span>
<span class="go">                    \edge.3--|</span>
<span class="go">                              \-Gorilla</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">quick_tree()</span></code> method also supports non-parametric bootstrapping. The number of resampled alignments is specified using the <code class="docutils literal notranslate"><span class="pre">bootstrap</span></code> argument. In the following, trees are estimated from 100 resampled alignments and merged into a single consensus topology using a weighted consensus tree algorithm.</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">tree</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">quick_tree</span><span class="p">(</span><span class="n">calc</span><span class="o">=</span><span class="s2">&quot;TN93&quot;</span><span class="p">,</span> <span class="n">bootstrap</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
</pre></div>
</div>
<div class="section" id="using-the-distancematrix-object">
<h3>Using the <code class="docutils literal notranslate"><span class="pre">DistanceMatrix</span></code> object<a class="headerlink" href="#using-the-distancematrix-object" title="Permalink to this headline">¶</a></h3>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_aligned_seqs</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aln</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span><span class="s1">&#39;data/primate_brca1.fasta&#39;</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dists</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">distance_matrix</span><span class="p">(</span><span class="n">calc</span><span class="o">=</span><span class="s2">&quot;TN93&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tree</span> <span class="o">=</span> <span class="n">dists</span><span class="o">.</span><span class="n">quick_tree</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tree</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">balanced</span><span class="p">()</span>  <span class="c1"># purely for display</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">ascii_art</span><span class="p">())</span>  
<span class="go">                    /-Rhesus</span>
<span class="go">          /edge.1--|</span>
<span class="go">         |         |          /-HowlerMon</span>
<span class="go">         |          \edge.0--|</span>
<span class="go">         |                    \-Galago</span>
<span class="go">-root----|</span>
<span class="go">         |--Orangutan</span>
<span class="go">         |</span>
<span class="go">         |          /-Chimpanzee</span>
<span class="go">          \edge.2--|</span>
<span class="go">                   |          /-Human</span>
<span class="go">                    \edge.3--|</span>
<span class="go">                              \-Gorilla</span>
</pre></div>
</div>
</div>
<div class="section" id="explicitly-via-distancematrix-and-cogent3-phylo-nj-nj">
<h3>Explicitly via <code class="docutils literal notranslate"><span class="pre">DistanceMatrix</span></code> and <code class="docutils literal notranslate"><span class="pre">cogent3.phylo.nj.nj()`</span></code><a class="headerlink" href="#explicitly-via-distancematrix-and-cogent3-phylo-nj-nj" title="Permalink to this headline">¶</a></h3>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent3.phylo</span> <span class="kn">import</span> <span class="n">nj</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_aligned_seqs</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aln</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span><span class="s1">&#39;data/primate_brca1.fasta&#39;</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dists</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">distance_matrix</span><span class="p">(</span><span class="n">calc</span><span class="o">=</span><span class="s2">&quot;TN93&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tree</span> <span class="o">=</span> <span class="n">nj</span><span class="o">.</span><span class="n">nj</span><span class="p">(</span><span class="n">dists</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tree</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">balanced</span><span class="p">()</span>  <span class="c1"># purely for display</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">ascii_art</span><span class="p">())</span>  
<span class="go">                    /-Rhesus</span>
<span class="go">          /edge.1--|</span>
<span class="go">         |         |          /-HowlerMon</span>
<span class="go">         |          \edge.0--|</span>
<span class="go">         |                    \-Galago</span>
<span class="go">-root----|</span>
<span class="go">         |--Orangutan</span>
<span class="go">         |</span>
<span class="go">         |          /-Chimpanzee</span>
<span class="go">          \edge.2--|</span>
<span class="go">                   |          /-Human</span>
<span class="go">                    \edge.3--|</span>
<span class="go">                              \-Gorilla</span>
</pre></div>
</div>
</div>
<div class="section" id="directly-from-a-pairwise-distance-dict">
<h3>Directly from a pairwise distance <code class="docutils literal notranslate"><span class="pre">dict</span></code><a class="headerlink" href="#directly-from-a-pairwise-distance-dict" title="Permalink to this headline">¶</a></h3>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent3.phylo</span> <span class="kn">import</span> <span class="n">nj</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dists</span> <span class="o">=</span> <span class="p">{(</span><span class="s1">&#39;a&#39;</span><span class="p">,</span> <span class="s1">&#39;b&#39;</span><span class="p">):</span> <span class="mf">2.7</span><span class="p">,</span> <span class="p">(</span><span class="s1">&#39;c&#39;</span><span class="p">,</span> <span class="s1">&#39;b&#39;</span><span class="p">):</span> <span class="mf">2.33</span><span class="p">,</span> <span class="p">(</span><span class="s1">&#39;c&#39;</span><span class="p">,</span> <span class="s1">&#39;a&#39;</span><span class="p">):</span> <span class="mf">0.73</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tree</span> <span class="o">=</span> <span class="n">nj</span><span class="o">.</span><span class="n">nj</span><span class="p">(</span><span class="n">dists</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">ascii_art</span><span class="p">())</span>  
<span class="go">          /-a</span>
<span class="go">         |</span>
<span class="go">-root----|--b</span>
<span class="go">         |</span>
<span class="go">          \-c</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="by-least-squares">
<h2>By Least-squares<a class="headerlink" href="#by-least-squares" title="Permalink to this headline">¶</a></h2>
<p>We illustrate the phylogeny reconstruction by least-squares using the F81 substitution model. We use the advanced-stepwise addition algorithm to search tree space. Here <code class="docutils literal notranslate"><span class="pre">a</span></code> is the number of taxa to exhaustively evaluate all possible phylogenies for. Successive taxa are added to the top <code class="docutils literal notranslate"><span class="pre">k</span></code> trees (measured by the least-squares metric) and <code class="docutils literal notranslate"><span class="pre">k</span></code> trees are kept at each iteration.</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent3.util.deserialise</span> <span class="kn">import</span> <span class="n">deserialise_object</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent3.phylo.least_squares</span> <span class="kn">import</span> <span class="n">WLS</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dists</span> <span class="o">=</span> <span class="n">deserialise_object</span><span class="p">(</span><span class="s1">&#39;data/dists_for_phylo.json&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ls</span> <span class="o">=</span> <span class="n">WLS</span><span class="p">(</span><span class="n">dists</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stat</span><span class="p">,</span> <span class="n">tree</span> <span class="o">=</span> <span class="n">ls</span><span class="o">.</span><span class="n">trex</span><span class="p">(</span><span class="n">a</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">show_progress</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
<p>Other optional arguments that can be passed to the <code class="docutils literal notranslate"><span class="pre">trex</span></code> method are: <code class="docutils literal notranslate"><span class="pre">return_all</span></code>, whether the <code class="docutils literal notranslate"><span class="pre">k</span></code> best trees at the final step are returned as a <code class="docutils literal notranslate"><span class="pre">ScoredTreeCollection</span></code> object; <code class="docutils literal notranslate"><span class="pre">order</span></code>, a series of tip names whose order defines the sequence in which tips will be added during tree building (this allows the user to randomise the input order).</p>
</div>
<div class="section" id="by-ml">
<h2>By ML<a class="headerlink" href="#by-ml" title="Permalink to this headline">¶</a></h2>
<p>We illustrate the phylogeny reconstruction using maximum-likelihood using the F81 substitution model. We use the advanced-stepwise addition algorithm to search tree space.</p>
<div class="highlight-pycon3 notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_aligned_seqs</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent3.phylo.maximum_likelihood</span> <span class="kn">import</span> <span class="n">ML</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">cogent3.evolve.models</span> <span class="kn">import</span> <span class="n">F81</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aln</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span><span class="s1">&#39;data/primate_brca1.fasta&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ml</span> <span class="o">=</span> <span class="n">ML</span><span class="p">(</span><span class="n">F81</span><span class="p">(),</span> <span class="n">aln</span><span class="p">)</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">ML</span></code> object also has the <code class="docutils literal notranslate"><span class="pre">trex</span></code> method and this can be used in the same way as for above, i.e. <code class="docutils literal notranslate"><span class="pre">ml.trex()</span></code>. We don’t do that here because this is a very slow method for phylogenetic reconstruction.</p>
</div>
</div>


          </div>
          
        </div>
      </div>
    <div class="clearer"></div>
  </div>
    <div class="footer">
      &copy;2020, cogent3.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 3.0.2</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="../_sources/cookbook/building_phylogenies.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>